### DISCUSSION TITLE
[Bug] python agent causes high cpu

### DISCUSSION THREAD

**Discussion by zzzzl13** [apache/skywalking#10642] [Q&A] [answered]
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/skywalking/issues?q=is%3Aissue) and found no similar issues.

### Apache SkyWalking Component

Python Agent (apache/skywalking-python)

### What happened

I have a python project with flask and use skywalking python agent. However when I enable the agent, the cpu will go high, even try disable the profile and log reporter, and even create a hello world flask, it still high cpu.
```
from flask import Flask
from skywalking import config, agent
import socket

app=Flask(__name__)

@app.route("/")
def index():
    return "<h1>Hello World!</h1>"
@app.route("/user/<name>")
def user(name):
    return "<h1>Hello,%s</h1>"%name
if __name__=="__main__":
    skw_host="skywalking_oap_host"
    skw_agent="local_test"
    config.init(agent_collector_backend_services="%s:11800"%skw_host, agent_name=skw_agent, agent_instance_name=socket.gethostname(), agent_protocol="grpc", agent_meter_reporter_period=60, agent_collector_heartbeat_period=60, agent_logging_level="ERROR",agent_profile_active=False, agent_log_reporter_active=False)
    agent.start()
    app.run(host="0.0.0.0",port=8000)
```
and use "python3.7 flask_test.py" to run this code

after "top -H", the cpu goes to 99.7%
 26 node      20   0  823184  62048  19744 R 99.7  1.5   8:41.35 python3.7

### What you expected to happen

I think it should not use that much cpu

### How to reproduce

rerun the code above

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

**Comment by wu-sheng** [comment]
I am moving this to the discussion. AFAIK, the community doesn't declare any performance payload metrics. So, nothing is about the bug.

**Comment by Superskyyy** [comment]
Hello, please provide your specs, dependency versions, agent version and how you integrated the agent. We noticed that few users face same problem recently and are investigating.

**Reply by zzzzl13** [reply]
python3.7.4
dependency：
  flask: 1.1.2
  apache-skywalking: 1.0.0

skywalking:9.4.0
and I just add this in code to start the skywalking agent
Thanks
```
    skw_host="skywalking_oap_host"
    skw_agent="local_test"
    config.init(agent_collector_backend_services="%s:11800"%skw_host, agent_name=skw_agent, agent_instance_name=socket.gethostname(), agent_protocol="grpc", agent_meter_reporter_period=60, agent_collector_heartbeat_period=60, agent_logging_level="ERROR",agent_profile_active=False, agent_log_reporter_active=False)
    agent.start()
    app.run(host="0.0.0.0",port=8000)
```

**Reply by alidisi** [reply]
python 3.7.9 
dependency 
  tornado: 6.1
  apache-skywalking: 1.0.0

skywalking agent start :
 config.init(agent_collector_backend_services=sky_addr, agent_name=agent_name, agent_protocol='http', agent_experimental_fork_support=True, agent_log_reporter_active=True, agent_meter_reporter_active=True, agent_profile_active=True)

![8007c2eaa3a5fc5cd98ecd96a96c1328](https://user-images.githubusercontent.com/19379933/230753600-0709b0f2-f5c2-4667-844b-3bfd75d520aa.png)
 cpu 100% 
strace  -p 24266
# look like futex(something like read/write socket） ,let cpu use rate hight
![50f62d13a0b66994fe3e9c853490b5d7](https://user-images.githubusercontent.com/19379933/230753609-e9f3ae54-9ee0-449e-b279-00f1bae026a6.png)

**Reply by Superskyyy** [reply]
@alidisi when you only turn on the trace reporter, does the cpu usage spike happen nonetheless?

**Reply by alidisi** [reply]
config.init(agent_collector_backend_services=sky_addr, agent_name=agent_name, agent_protocol='http', agent_experimental_fork_support=True, agent_log_reporter_active=False, agent_meter_reporter_active=False, agent_profile_active=False) ;
 i try close agent_log_reporter_active/agent_meter_reporter_active/agent_profile_active .cpu usage  rate go down a little , but still very hight
<img width="627" alt="image" src="https://user-images.githubusercontent.com/19379933/231049090-718ea6f7-0137-48ae-bf75-4e60f32d03a5.png">

**Reply by Superskyyy** [reply]
@alidisi If you have time, could you profile the agent-enabled application with [Yappi](https://github.com/sumerc/yappi)? So we can see clearly what exact line is taking the CPU time.

**Reply by Superskyyy** [reply]
> @alidisi If you have time, could you profile the agent-enabled application with [Yappi](https://github.com/sumerc/yappi)? So we can see clearly what exact line is taking the CPU time. 

or i will do it over this weekend. After fixing it a new release will be published immediately.

**Reply by zzzzl13** [reply]
Not sure if it is like this
![2023-04-11 14_04_48-Window](https://user-images.githubusercontent.com/61879466/231071799-be834e76-9852-4570-a038-a98f2e62afe2.png)

**Comment by FAWC438** [comment]
I seem to have found where the problem is. 

[These codes](https://github.com/apache/skywalking-python/blob/128278a1d61b6a92e67ee9c67cc2836a91370bbc/skywalking/agent/__init__.py#L310-L318) in  agent/\_\_init\_\_.py cause the bug.

These codes results in a timeout waiting time of **0** forever when fetching the in-queue elements of the segment and log, which means that the thread has been hogging the CPU without any time to “Rest” it. Meanwhile, the `backoff_decorator` uses `threading.Event().wait()` , which in Python 3 is implemented in C at the bottom, means that it is not restricted by GIL, resulting in multi-core CPU usage easily exceeding 100.0.

To solve this bug , just set the parameter `init_wait` in the decorator of the two functions to a small non-zero positive number (as shown below, I set it to 0.01).

```python
    @report_with_backoff(reporter_name='segment', init_wait=0.01)
    def __report_segment(self) -> None:
        if not self.__segment_queue.empty():
            self.__protocol.report_segment(self.__segment_queue)

    @report_with_backoff(reporter_name='log', init_wait=0.01)
    def __report_log(self) -> None:
        if not self.__log_queue.empty():
            self.__protocol.report_log(self.__log_queue)
```

The following is a comparison of CPU usage _before_ and _after_ making this change:

_before_

```bash
   PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND                                                                                                                                                             
184153 root      20   0 2400796 136480  18488 S 115.0  0.2   0:18.67 python 
```


_after_

```bash

   PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND                                                                                                                                                             
181723 root      20   0 2475044 137004  18392 S   3.0  0.2   0:07.43 python
```

**Reply by Superskyyy** [reply]
Through the yappi profile result I see this is true as event wait hogs all cpu time. Although it was already implemented as wait(0) in 0.8.0, and it wasn't breaking back then.

**Comment by Superskyyy** [comment, marked as answer]
New tracking issue created, pending a new release once this is fixed and merged into master.

**Comment by Mushano** [comment]
It seems that I also encountered this problem, after using the Python agent version 1.0 probe CPU load pulling full. Stable at 20% CPU usage with version 0.8.

**Reply by Superskyyy** [reply]
> It seems that I also encountered this problem, after using the Python agent version 1.0 probe CPU load pulling full. Stable at 20% CPU usage with version 0.8.

Thanks for reporting and sorry for the regression, a new release v1.0.1 will be released soon. Workaround at your own environment (if urgent) is to change the agent/__init__.py line 310 and 315 init_wait parameter to 0.01 instead of current 0.

**Comment by Superskyyy** [comment]
@zzzzl13 @Mushano A fix is merged, please feel free to try it out in master branch or wait for an official release (v1.0.1) within three days.

**Reply by zzzzl13** [reply]
OK, thanks
